### PR TITLE
Port to Gen.jl style interfaces

### PR DESCRIPTION
**Pull request by sritchie**
This PR:

- Implements a new `ChoiceMap` and `Trace` type for the dynamic language that matches Gen.jl more closely
  - These use the new interfaces in `gen.choicemap` and `gen.trace`
  - `DynamicDSLFunction` gains `gf/IAssess` and `gf/IPropose` implementations`
- Upgrades the Clerk notebooks to use the new code
- Modifies `#gen/choicemap` and `#gen/choice` to produce instances from `gen.choicemap`
- `In gen.distribution`:
  - `GenerativeFn` now implements `accepts-output-grad?` and `get-params`, and the methods for `gf/IAssess` and `gf/IPropose`
  - `GenerativeFn` also gains a new `arity` field, which is used to implement `has-argument-grads`
- All of the `commons-math`, `kixi` and `java-util` distributions use this new constructor
- `gen.trace/update` and `gen.generative-function/generate` are now more permissive and can accept normal maps for constraints

### CODE REVIEW

File: src/gen/distribution.cljc [hunk 1]
```diff
@@ -39,22 +109,46 @@
 ;;
 ;; This type provides support for all primitive distributions.
 
-(defrecord GenerativeFn [ctor]
+(defrecord GenerativeFn [ctor arity]
```
Comment (sritchie) [inline src/gen/distribution.cljc:112]: this is required to support `has-argument-grads` below (well, to support declaring that this is not supported yet!)

File: src/gen/distribution.cljc [hunk 2]
```diff
@@ -27,6 +29,74 @@
   (and (satisfies? LogPDF t)
        (satisfies? Sample t)))
 
+;; ## Combinators
```
Comment (sritchie) [inline src/gen/distribution.cljc:32]: these moved up from below.

**Review by zane — APPROVED** [review]
Congratulations on getting this PR over the finish line! I'm very excited about Gen.clj's future.

This is not my most thorough review; I think I'm missing some of the context and this is a pretty hefty PR. But I left some thoughts, suggestions, and questions inline.

File: test/gen/distribution_test.cljc [hunk 3]
```diff
@@ -168,13 +186,13 @@
...
              sigma (gen-double 0.001 10)
-             v     (gen-double -100 100)
+             v     (gen-double -50 50)
              shift (gen-double -10 10)]
             (is (ish? (dist/logpdf (->normal 0.0 sigma) v)
                       (dist/logpdf (->normal 0.0 sigma) (- v)))
                 "Normal is symmetric about the mean")
 
-            (with-comparator (within 1e-12)
+            (with-comparator (within 1e-10)
```
[the remark this answers is not in the document]
Comment (sritchie) [reply test/gen/distribution_test.cljc:195]: These tests have been flaky, so I'm relaxing the bounds a touch.

**Pull request merged**